address/address.go

```diff
@@ -47,327 +68,267 @@ const (
+	return Address{string(buf)}, nil
+}
 
-	for i, p := range toEncode {
-		if p >= 32 {
-			return "", fmt.Errorf("invalid data")
-		}
-		encoded[i] += Base32Charset[p]
+func encode(network Network, addr Address) (string, error) {
+	if addr == Undef {
+		return UndefAddressString, nil
+	}
+	var ntwk string
+	switch network {
+	case Mainnet:
+		ntwk = MainnetPrefix
+	case Testnet:
+		ntwk = TestnetPrefix
+	default:
+		return UndefAddressString, ErrUnknownNetwork
 	}
 
-	return hrp + string(encoded), nil
+	var strAddr string
+	switch addr.Protocol() {
+	case SECP256K1, Actor, BLS:
+		cksm := Checksum(append([]byte{addr.Protocol()}, addr.Payload()...))
+		strAddr = ntwk + fmt.Sprintf("%d", addr.Protocol()) + AddressEncoding.WithPadding(-1).EncodeToString(append(addr.Payload(), cksm[:]...))
+	case ID:
+		strAddr = ntwk + fmt.Sprintf("%d", addr.Protocol()) + fmt.Sprintf("%d", leb128.ToUInt64(addr.Payload()))
+	default:
+		return UndefAddressString, ErrUnknownProtocol
+	}
+	return strAddr, nil
 }
 
-// decode decodes a filecoin address and returns
-// the hrp(human-readable part), version (1byte) and data(32bit data array)
-// or an error.
-func decode(addr string) (string, byte, []byte, error) {
-	if len(addr) < 2 {
-		return "", 0, nil, fmt.Errorf("address too short")
+func decode(a string) (Address, error) {
+	if len(a) == 0 {
+		return Undef, nil
 	}
-	if len(addr) > 41 {
-		return "", 0, nil, fmt.Errorf("too long: len=%d", len(addr))
+	if a == UndefAddressString {
+		return Undef, nil
 	}
-	if strings.ToLower(addr) != addr && strings.ToUpper(addr) != addr {
-		return "", 0, nil, fmt.Errorf("mixed case")
-	}
-	addr = strings.ToLower(addr)
-	hrp := addr[0:2]
-	for p, c := range hrp {
-		if c < 33 || c > 126 {
-			return "", 0, nil, fmt.Errorf("invalid character human-readable part: addr[%d]=%d", p, c)
-		}
+	if len(a) > MaxAddressStringLength || len(a) < 3 {
+		return Undef, ErrInvalidLength
 	}
 
-	// check data
-	data := addr[2:]
-	dataBytes := []byte{}
-	for _, b := range data {
-		// alphanumeric only
-		if !((b >= int32('0') && b <= int32('9')) || (b >= int32('A') && b <= int32('Z')) || (b >= int32('a') && b <= int32('z'))) {
-			return "", 0, nil, fmt.Errorf("non alphanumeric character")
-		}
-
-		// exclude [1,b,i,o]
-		if b == int32('1') || b == int32('b') || b == int32('i') || b == int32('o') {
-			return "", 0, nil, fmt.Errorf("invalid character")
-		}
+	if string(a[0]) != MainnetPrefix && string(a[0]) != TestnetPrefix {
+		return Undef, ErrUnknownNetwork
+	}
 
-		dataBytes = append(dataBytes, byte(Base32CharsetReverse[int(b)]))
+	var protocol Protocol
+	switch a[1] {
+	case '0':
+		protocol = ID
+	case '1':
+		protocol = SECP256K1
+	case '2':
+		protocol = Actor
+	case '3':
+		protocol = BLS
+	default:
+		return Undef, ErrUnknownProtocol
 	}
 
-	if !verifyChecksum(hrp, dataBytes) {
-		return "", 0, nil, fmt.Errorf("invalid checksum")
+	raw := a[2:]
+	if protocol == ID {
+		id, err := strconv.ParseUint(raw, 10, 64)
```

Comment:
I will add a check specifically for `len(raw) > 20` (since 20 is length of math.MaxUint64 as a string), although I believe the `strconv.ParseUint` check handles this for us.